x/subscription/spec/06_begin_block.md

```diff
@@ -0,0 +1,56 @@
+<!--
+order: 6
+-->
+
+# BeginBlock
+
+## Remove expired plans
+
+```python
+for plan_id, expiration_time in get_plans_sorted_by_expiration_time():
+    if block_time >= expiration_time:
+        stop_plan(plan_id)
+    else:
+        break
+```
+
+## Remove expired subscriptions
+
+```python
+for plan_id, subscriber, expiration_time in get_subscriptions_sorted_by_expiration_time():
+    if block_time >= expiration_time:
+        stop_subscription(plan_id, subscriber)
+    else:
+        break
+```
+
+## Collect payments
+
+`x/subscription` module automatically collect payments from subscribers at begin block, and write collection results into events.
+
+```python
+for plan_id, next_collectime_time in get_plans_sorted_by_next_collection_time():
```

Comment:
what is next collection time? perhaps iterate subscriptions directly?

Reply:
yeah, it seems makes more sense to iterate subscriptions directly, and in the new revision, failed subscription get retried automatically. I also makes the `FailureTolerance` to be a network parameter. failed subscription get stopped after retried some times.

Reply:
the cancellation option may still need to be there -- e.g. if the plan owner terminates the service for particular/selected subscribers (while still keeping the other subscriptions going), they may also want to proactively cancel their subscriptions regardless of their balance, so they don't get charged by mistake (as affected subscribers may forget to unsubscribe / cancel themselves)